packages/aws-rfdk/lib/deadline/lib/database-connection.ts

```diff
@@ -175,7 +181,12 @@ class DocDBDatabaseConnection extends DatabaseConnection {
    */
   public readonly containerEnvironment: { [name: string]: string };
 
-  constructor(private readonly props: DocDBConnectionOptions) {
+  /**
+   * @inheritdoc
+   */
+  public readonly database: Construct;
+
+  constructor(scope: Construct, private readonly props: DocDBConnectionOptions) {
```

Comment:
This is a breaking API change. We should already have a reference to the stack of the database from `props.database`. Technically, it has a type of `IDatabase`, but can we use type-checking at run-time to determine whether this is the L2 construct vs the imported resource? So I'm thinking something like:

```ts
if (props.database instanceof DatabaseCluster) {
    // props.database is a L2 DatabaseCluster
    // ...
} else {
    // props.database is an imported DatabaseCluster
    // ...
}
```

what do you think?

Reply:
the problem is that `props.database` is a `DatabaseConnection` and not a `DatabaseCluster`. There are 2 implementations of `DatabaseConnection` and neither expose the construct that represents the actual database.

1. `DocDBDatabaseConnection` stores its `props` as a private field and the `IDatabaseCluster` is the `database` field on that object.
2. `MongoDbInstanceDatabaseConnection` also stores its `props` as a private field, but the `database` field on that object is an `IMongoDb`.

What if we add the abstract method `isDatabaseImported` to `DatabaseConnection` that just returns true or false? I think the Mongo implementation will always return false, but the DocDB version would do the `instanceof` test to see if it's a class or interface.

Reply:
An alternative to my suggestion is the visitor pattern. It would be more extensible going forward, but it's a bit heavy-duty.

Reply:
Thank you for advices. What about this version? Does it look better?

Reply:
You're not traversing down into the MongoDbInstanceConnection properly to determine if it's using an import or not, but otherwise this looks like it should work. However, I think the new public class variable to hold the database construct could be confusing to anybody using the DatabaseConnection class; it isn't obvious why it's there or why it might not be populated with anything. I guess if we documented it properly then it should be fine.


Determining if the MongoDB connection is an import might be something like this:
```
if (mongoDbConnection.database instanceof MongoDbInstance) {
  dbInstance = (MongoDbInstance) mongoDbConnection.database;
  if (dbInstance.mongoDataVolume instanceof Volume) {
    return dbInstance.mongoDataVolume;
  }
}
```

Reply:
Looks like we've pivoted to this, so I'll resolve this one.